File: app/src/main/java/com/ivanmagda/habito/utils/HabitoDateUtils.java
package com.ivanmagda.habito.utils;

import android.text.format.DateUtils;

import com.ivanmagda.habito.models.ResetFrequency;

import java.util.Calendar;
import java.util.Date;

public final class HabitoDateUtils {

    public static Calendar getCurrentCalendar() {
        return getCalendarWithTime(System.currentTimeMillis());
    }

    public static Calendar getCalendarWithTime(long time) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(time);
        return calendar;
    }

    public static boolean isWithinRange(long toCheck, long start, long end) {
        return isWithinRange(new Date(toCheck), new Date(start), new Date(end));
    }

    public static boolean isWithinRange(Date toCheck, Date start, Date end) {
        return !(toCheck.before(start) || toCheck.after(end));
    }

    public static boolean isDateInCurrentWeek(long date) {
        Calendar currentCalendar = getCurrentCalendar();
        int week = currentCalendar.get(Calendar.WEEK_OF_YEAR);
        int year = currentCalendar.get(Calendar.YEAR);

        Calendar targetCalendar = getCalendarWithTime(date);
        int targetWeek = targetCalendar.get(Calendar.WEEK_OF_YEAR);
        int targetYear = targetCalendar.get(Calendar.YEAR);

        return (week == targetWeek) && (year == targetYear);
    }

    public static boolean isDateInCurrentMonth(long date) {
        Calendar currentCalendar = getCurrentCalendar();
        int month = currentCalendar.get(Calendar.MONTH);
        int year = currentCalendar.get(Calendar.YEAR);

        Calendar targetCalendar = getCalendarWithTime(date);
        int targetMonth = targetCalendar.get(Calendar.MONTH);
        int targetYear = targetCalendar.get(Calendar.YEAR);

        return (month == targetMonth) && (year == targetYear);
    }

    public static boolean isDateInCurrentYear(long date) {
        Calendar currentCalendar = getCurrentCalendar();
        int year = currentCalendar.get(Calendar.YEAR);

        Calendar targetCalendar = getCalendarWithTime(date);
        int targetYear = targetCalendar.get(Calendar.YEAR);

        return (year == targetYear);
    }

    public static boolean isDateInType(long date, ResetFrequency.Type type) {
        switch (type) {
            case DAY:
                return DateUtils.isToday(date);
            case WEEK:
                return HabitoDateUtils.isDateInCurrentWeek(date);
            case MONTH:
                return HabitoDateUtils.isDateInCurrentMonth(date);
            case YEAR:
                return HabitoDateUtils.isDateInCurrentYear(date);
            case NEVER:
                return true;
            default:
                throw new IllegalArgumentException("Received illegal type");
        }
    }

    public static long getStartOfThisWeek() {
        Calendar calendar = getCurrentCalendar();
        calendar.set(Calendar.HOUR_OF_DAY, 0);
        calendar.clear(Calendar.MINUTE);
        calendar.clear(Calendar.SECOND);
        calendar.clear(Calendar.MILLISECOND);

        // get start of this week in milliseconds
        calendar.set(Calendar.DAY_OF_WEEK, calendar.getFirstDayOfWeek());

        return calendar.getTimeInMillis();
    }

    public static boolean sameDay(long lhs, long rhs) {
        Calendar lhsCal = Calendar.getInstance();
        Calendar rhsCal = Calendar.getInstance();
        lhsCal.setTimeInMillis(lhs);
        rhsCal.setTimeInMillis(rhs);
        return (lhsCal.get(Calendar.YEAR) == rhsCal.get(Calendar.YEAR))
                && (lhsCal.get(Calendar.DAY_OF_YEAR) == rhsCal.get(Calendar.DAY_OF_YEAR));
    }

    public static int getNumberOfWeeksInCurrentMonth() {
        Calendar calendar = getCalendarWithTime(getStartOfCurrentMonth());
        return calendar.getActualMaximum(Calendar.WEEK_OF_MONTH);
    }

    public static long getStartOfCurrentMonth() {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        calendar.set(Calendar.DAY_OF_MONTH, 1);
        return calendar.getTimeInMillis();
    }

    public static long getEndOfCurrentMonth() {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        calendar.set(Calendar.DAY_OF_MONTH, calendar.getActualMaximum(Calendar.DAY_OF_MONTH));
        return calendar.getTimeInMillis();
    }

    private HabitoDateUtils() {
    }

}

File: app/src/main/java/com/ivanmagda/habito/utils/HabitoDateUtils.kt
package com.ivanmagda.habito.utils

import android.text.format.DateUtils
import com.ivanmagda.habito.models.ResetFrequency
import java.util.*

object HabitoDateUtils {

    val currentCalendar: Calendar
        get() = getCalendarWithTime(System.currentTimeMillis())

    val startOfCurrentWeek: Long
        get() = getStartOfWeek(System.currentTimeMillis())

    val endOfCurrentWeek: Long
        get() {
            val calendar = getCalendarWithTime(startOfCurrentWeek)
            calendar.add(Calendar.DATE, 6)

            return calendar.timeInMillis
        }

    val numberOfWeeksInCurrentMonth: Int
        get() {
            val calendar = getCalendarWithTime(startOfCurrentMonth)

            return calendar.getActualMaximum(Calendar.WEEK_OF_MONTH)
        }

    val startOfCurrentMonth: Long
        get() {
            val calendar = currentCalendar
            calendar.set(Calendar.DAY_OF_MONTH, 1)

            return calendar.timeInMillis
        }

    val endOfCurrentMonth: Long
        get() {
            val calendar = currentCalendar
            calendar.set(Calendar.DAY_OF_MONTH, calendar.getActualMaximum(Calendar.DAY_OF_MONTH))

            return calendar.timeInMillis
        }

    val startOfCurrentYear: Long
        get() {
            val calendar = currentCalendar
            calendar.set(Calendar.MONTH, 0)

            return calendar.timeInMillis
        }

    val endOfCurrentYear: Long
        get() {
            val calendar = currentCalendar
            calendar.set(Calendar.MONTH, 11)

            return calendar.timeInMillis
        }

    fun getCalendarWithTime(time: Long): Calendar {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = time

        return calendar
    }

    fun isWithinRange(toCheck: Long, start: Long, end: Long): Boolean {
        return isWithinRange(Date(toCheck), Date(start), Date(end))
    }

    fun isWithinRange(toCheck: Date, start: Date, end: Date): Boolean {
        return !(toCheck.before(start) || toCheck.after(end))
    }

    fun isDateInCurrentWeek(date: Long): Boolean {
        return isDatesInSameMonth(currentCalendar.timeInMillis, getCalendarWithTime(date).timeInMillis)
    }

    fun isDatesInSameWeek(lhsDate: Long, rhsDate: Long): Boolean {
        val lhsCalendar = getCalendarWithTime(lhsDate)
        val lhsWeek = lhsCalendar.get(Calendar.WEEK_OF_YEAR)
        val lhsYear = lhsCalendar.get(Calendar.YEAR)

        val rhsCalendar = getCalendarWithTime(rhsDate)
        val rhsWeek = rhsCalendar.get(Calendar.WEEK_OF_YEAR)
        val rhsYear = rhsCalendar.get(Calendar.YEAR)

        return lhsWeek == rhsWeek && lhsYear == rhsYear
    }

    fun isDateInCurrentMonth(date: Long): Boolean {
        return isDatesInSameMonth(currentCalendar.timeInMillis, getCalendarWithTime(date).timeInMillis)
    }

    fun isDatesInSameMonth(lhs: Long, rhs: Long): Boolean {
        val lhsCalendar = getCalendarWithTime(lhs)
        val lhsMonth = lhsCalendar.get(Calendar.MONTH)
        val lhsYear = lhsCalendar.get(Calendar.YEAR)

        val rhsCalendar = getCalendarWithTime(rhs)
        val rhsMonth = rhsCalendar.get(Calendar.MONTH)
        val rhsYear = rhsCalendar.get(Calendar.YEAR)

        return lhsMonth == rhsMonth && lhsYear == rhsYear
    }

    fun isDateInCurrentYear(date: Long): Boolean {
        val currentCalendar = currentCalendar
        val year = currentCalendar.get(Calendar.YEAR)

        val targetCalendar = getCalendarWithTime(date)
        val targetYear = targetCalendar.get(Calendar.YEAR)

        return year == targetYear
    }

    fun isDateInType(date: Long, type: ResetFrequency.Type): Boolean {
        return when (type) {
            ResetFrequency.Type.DAY -> DateUtils.isToday(date)
            ResetFrequency.Type.WEEK -> HabitoDateUtils.isDateInCurrentWeek(date)
            ResetFrequency.Type.MONTH -> HabitoDateUtils.isDateInCurrentMonth(date)
            ResetFrequency.Type.YEAR -> HabitoDateUtils.isDateInCurrentYear(date)
            ResetFrequency.Type.NEVER -> true
        }
    }

    fun getStartOfWeek(week: Long): Long {
        val calendar = getCalendarWithTime(week)
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.clear(Calendar.MINUTE)
        calendar.clear(Calendar.SECOND)
        calendar.clear(Calendar.MILLISECOND)

        // get start of this week in milliseconds
        calendar.set(Calendar.DAY_OF_WEEK, calendar.firstDayOfWeek)

        return calendar.timeInMillis
    }

    fun isSameDay(lhs: Long, rhs: Long): Boolean {
        val lhsCal = Calendar.getInstance()
        val rhsCal = Calendar.getInstance()

        lhsCal.timeInMillis = lhs
        rhsCal.timeInMillis = rhs

        return lhsCal.get(Calendar.YEAR) == rhsCal.get(Calendar.YEAR) &&
                lhsCal.get(Calendar.DAY_OF_YEAR) == rhsCal.get(Calendar.DAY_OF_YEAR)
    }
}
